edgedb/retry.py

```diff
@@ -0,0 +1,87 @@
+import asyncio
+import random
+
+from . import errors
+from . import transaction as _transaction
+
+
+DEFAULT_MAX_ITERATIONS = 3
+
+
+def default_backoff(attempt):
+    return (2 ** attempt) * 0.1 + random.randrange(100) * 0.1
+
+
+class AsyncIOIteration(_transaction.AsyncIOTransaction):
+    def __init__(self, retry, owner):
+        super().__init__(owner)
+        self.__retry = retry
+
+    async def start(self):
+        if not self._managed:
+            raise errors.InterfaceError(
+                "Only managed retriable transactions are supported. "
+                "Use `async with transaction:`"
+            )
+        # TODO(tailhook) if this is not the first iteration suppress
+        # `wait_until_available` timeout
+        await super().start()
+
+    async def __aexit__(self, extype, ex, tb):
+        try:
+            await super().__aexit__(extype, ex, tb)
+        except errors.ClientConnectionClosedError as err:
+            if ex is None:
+                # TODO(tailhook) should we figure out if err is *caused by*
+                # CancelledError and propagate the latter instead or retrying?
```

Comment:
Hm, *can* it be caused by CancelledError? CancelledError is a BaseException, it shouldn't be wrapped into ClientConnectionClosedError by us.

Reply:
Here is how it happens:
```
Traceback (most recent call last):
  File "/work/tests/test_async_retry.py", line 96, in transaction1
    res = await tx.query_one('''
  File "/work/edgedb/transaction.py", line 238, in query_one
    result, _ = await self._connection_impl._protocol.execute_anonymous(
  File "edgedb/protocol/protocol.pyx", line 622, in execute_anonymous
    ret, attrs = await self._execute(in_dc, out_dc, args, kwargs)
  File "edgedb/protocol/protocol.pyx", line 362, in _execute
    await self.wait_for_message()
  File "edgedb/protocol/asyncio_proto.pyx", line 70, in wait_for_message
    raise
  File "edgedb/protocol/asyncio_proto.pyx", line 59, in edgedb.protocol.asyncio_proto.AsyncIOProtocol.wait_for_message
    await self.msg_waiter
asyncio.exceptions.CancelledError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/work/tests/test_async_retry.py", line 108, in transaction1
    await barrier.ready()
  File "/work/edgedb/retry.py", line 34, in __aexit__
    await super().__aexit__(extype, ex, tb)
  File "/work/edgedb/transaction.py", line 155, in __aexit__
    await self.__rollback()
  File "/work/edgedb/transaction.py", line 199, in __rollback
    await self._connection_impl.execute(query)
  File "/work/edgedb/asyncio_con.py", line 153, in execute
    await self._protocol.simple_query(query)
  File "edgedb/protocol/protocol.pyx", line 533, in simple_query
    self.ensure_connected()
  File "edgedb/protocol/protocol.pyx", line 329, in edgedb.protocol.protocol.SansIOProtocol.ensure_connected
    raise errors.ClientConnectionClosedError(
edgedb.errors.ClientConnectionClosedError: the connection has been closed because an operation was cancelled on it
```
Basically it's because `__exit__` is called on `BaseExceptions` too.

But generally this code is non-controversial, regardless of the `CancelledError`:
1. I've changed it to capture any error

Reply:
On the second thought, this is where `EarlyNetworkError` needed to kick in. I.e. this is `COMMIT` branch if it failed, we have to be very careful when retrying transaction. So I need a separate tag here.

Reply:
Another round: It looks like serialization error happens on the actual query rather than commit. Since this is the most important thing we want to fix I'm simplifying the code (by raising all errors during commit and swallowing all errors during rollback), and keep it as is. We can improve robustness here later.